### PR TITLE
Brodes/guard flow parsing

### PR DESCRIPTION
**Pull request by bdrodes**

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 1]
```diff
@@ -5,9 +5,21 @@
...
 private import semmle.code.cpp.ir.implementation.raw.internal.TranslatedExpr
 private import semmle.code.cpp.ir.implementation.raw.internal.InstructionTag
 
+/**
+ * Returns `instr` or any instruction used to define `instr`.
+ */
+private Instruction getDerivedInstruction(Instruction instr) {
+  result = valueNumber(instr).getAnInstruction() and
+  result.toString() != instr.toString() and
+  result instanceof CompareInstruction
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:18]: We don't want to use `toString` when implementing query/library logic. `toString` output is being changed consistently to provide better UX in query output without any deprecation periods, and there shouldn't be any hidden invariants in the output of these predicates. So we need to come up with something else here.

What's the job of this predicate really? I _think_ you're trying to find an equivalent instruction that performs a comparison so that we can derive guards from that comparison instead of the original instruction, right? But why should we restrict that to be `CompareInstruction`s only? And why check that they have a different `toString`? If it's because we want to not have `result = instr` in this case could we not simply write `result != instr`?

And related to the QLDoc: What do you mean by "define" `instr`? I think we should have an example in the QLDoc to motivate this predicate.

Comment (bdrodes) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:18]: I'll take the discussion offline. But some quick comments:
- Agreed on tostring, but I don't know a better approach, we will discuss
- I can't say result != instr because that's not what I'm trying to achieve here. I'm trying to not get other GVN instructions that are identical. e.g., `a->foo->bar`  I don't need to get every a->foo->bar, I need to get the instruction that defines a->foo->bar and ignore everything else. I'm playing a game with GVN to give me the SSA def. We probably need to use SSA instead, but this was a quick demonstration of how to get what I wanted (it at least works with the AST version of this in my tests in the past)
- I only limited CompareInstruction because I thought I saw you were doing the same, but I've been testing with and without it. 
- The TLDR for me is either we use SSA, or keep this kind of approach but use the IR to help me find only the GVN with the definition. I'm no expert on the syntax of all the different IR instruction types, but perhaps you can provide some guidance. 

Taking further discussion into a DM until we settle on a few things.

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 2]
```diff
@@ -13,9 +13,18 @@ private import semmle.code.cpp.ir.implementation.raw.internal.InstructionTag
...
+  // The instruction itself
+  result = instr
+  or
+  // or the GVN definition for the current instruction
   result = valueNumber(instr).getAnInstruction().(StoreInstruction).getSourceValue()
   or
-  result = instr
+  // Special handling for conversions
+  result =
+    getDerivedInstruction(valueNumber(instr).getAnInstruction().(CompareNEInstruction).getLeft())
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:24]: You should probably also check that the right-hand side has an `int_value` of 0 to make sure this is _actually_ an int-to-bool conversion.

I also think this deserves a more descriptive comment than what is currently here.

Comment (bdrodes) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:24]: Can you explain that a bit? What does a value of 0 or non-zero mean?

**Comment by bdrodes** [comment]
Worked with @MathiasVP  on a prior solution of his to find inefficiencies and worked out conversion concerns. We are going to take the lessons learned here and put them into that PR. Closing this draft.
